pkg/init/interface.go

```diff
@@ -1,9 +1,40 @@
+// Package init provides methods to initiate an odo project.
+// Most of the methods of the package get a `flags` parameter
+// representing the flags passed from the user through the command line.
+// Several backends are available to complete the operations, the backend
+// being chosen depending on the flags content:
+// - if no flags are passed, the `interactive` backend will be used
+// - if some flags are passed, the `flags` backend will be used.
 package init
 
-import "github.com/devfile/library/pkg/devfile/parser"
+import (
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+
+	"github.com/redhat-developer/odo/pkg/init/backend"
+)
 
 type Client interface {
```

Comment:
This `Client` interface it seems implements the `InitBackend`. This seems confusing to me. For `InitClient`, all the methods in common check if the backend is a flags or interactive, and then call it's method. Why can't we assign a backend when initializing `InitClient` and remove these methods from `Client` interface?

Point being, can we make this work without defining the same method in both `Client` and `InitBackend`?

Reply:
I see @dharmit had raised the same question. IIUC, Dharmit is suggesting that we assign the backend while initializing the `InitClient`, so in the `NewInitClient` factory function?

Reply:
> This `Client` interface it seems implements the `InitBackend`. This seems confusing to me. For `InitClient`, all the methods in common check if the backend is a flags or interactive, and then call it's method. Why can't we assign a backend when initializing `InitClient` and remove these methods from `Client` interface?
> 

See https://github.com/redhat-developer/odo/pull/5434#issuecomment-1034560129

> Point being, can we make this work without defining the same method in both `Client` and `InitBackend`?

The `Client` and `InitBackend` interfaces are very similar, but this is by "chance", not "by design". If you look at the other implementation (before my last commit), the signatures of the methods were different. Coupling the interface of `Client` and `InitBackend` would limit the possibilities to adapt one without affecting the other.

Reply:
> IIUC, Dharmit is suggesting that we assign the backend while initializing the `InitClient`, so in the `NewInitClient` factory function?

Nope, not in the `NewInitClient` factory function. More like in `Complete` or `Validate` phase of CLI. Factory method is, in my understanding, definitely not the place to do this. But I agree with what @feloy said:

> Because this routine will be used in odo init, dev and deploy, we want to factorize the logic in a common package (pkg/init).